### ISSUE TITLE
False positives with Credo.Check.Refactor.Apply

### ISSUE THREAD

**Issue by maltoe** [rrrene/credo#980]
Hi :wave: 

We've run into two separate cases of "false positives" with `Credo.Check.Refactor.Apply` now, so I figured I might ask whether this check makes sense at all in its current implementation. Consider the following code:

## Case 1

```
def case1(some_dynamic_list) do
  apply(__MODULE__, :foo, [:some_value | some_dynamic_list])
end
```

This is obviously (see the variable name :upside_down_face:) a list of which we don't know the exact length beforehand, yet the wrapping in `[ | ]` to prepend the constant value makes Credo see the list and complain. Our actual use-case is a `path_helper` prop in some UI component, where we prepend the `conn` to an unknown path helper + given list of following parameters.

Can be fixed like this, but quite ugly so:

```elixir
def case1(some_dynamic_list) do
  args = [:some_value | some_dynamic_list]
  apply(__MODULE__, :foo, args)
end
```

## Case 2

In the second case, we do in fact know the number of parameters we want to pass to a function, but we don't know whether the function exists on the module. Changing this to `__MODULE__.bar()` creates a compiler warning in case the function is not defined.

```elixir
defmacro case2 do
  quote do
    def foo do
      if function_exported?(__MODULE__, :bar, 0) do
        apply(__MODULE__, :bar, [])
      end
    end
  end
end
```

Can be fixed with the same move of `[]` to a local variable. This one is also particularly annoying, as the `disable-for-next-line` comment or similar won't work due to the macro expansion.

## Conclusion

So essentially we have disabled the check entirely, and rely on conventional wisdom instead. If I have to use `apply/2` or `apply/3` for whatever reason, I probably know what I'm doing.

Question would be, does this check make sense at all? Could it be made smarter to handle these cases? Or are there perhaps alternative ways to express these cases?

Otherwise thank you for the amazing library! :purple_heart: 

Best,
malte

### Environment

* credo 1.6.4
* Elixir 1.13.2 on OTP 22

**Comment by NickNeck** [comment]
Good catch!
>  Does this check make sense at all? 

I have already found one or two places in my projects that I have refactored with this check. But the prio of the check is very low.

> Could it be made smarter to handle these cases? 

I have tried #982
